File: HackerRank/Dynamic_Programming/Sherlock_and_Cost.py
import sys

def cost(b):
    max_values = [[0]*2 for _ in range(len(b))]
    for i in range(1, len(b)):
        max_values[i][0] = max(max_values[i-1][0], max_values[i-1][1]+b[i-1]-1)
        max_values[i][1] = max(max_values[i-1][0]+b[i]-1, max_values[i-1][1]+abs(b[i-1]-b[i]))
    return max(max_values[len(b)-1][0], max_values[len(b)-1][1])

if __name__ == "__main__":
    t = int(sys.stdin.readline())
    for i in range(0, t):
        n = int(sys.stdin.readline())
        b = list(map(int, sys.stdin.readline().split()))
        print(cost(b))

File: HackerRank/Dynamic_Programming/Sherlock_and_Cost.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var t int
	fmt.Fscanln(reader, &t)

	for i := 0; i < t; i++ {
		var n int
		fmt.Fscanln(reader, &n)
		var b = make([]int, n)
		for j := 0; j < n; j++ {
			fmt.Fscanf(reader, "%d ", &b[j])
		}
		fmt.Fprintln(writer, cost(b))
	}
}

func cost(b []int) int {
	var maxValues = make([][]int, len(b))
	for i := 0; i < len(maxValues); i++ {
		maxValues[i] = make([]int, 2)
	}

	for i := 1; i < len(b); i++ {
		maxValues[i][0] = int(math.Max(float64(maxValues[i-1][0]), float64(maxValues[i-1][1]+b[i-1]-1)))
		maxValues[i][1] = int(math.Max(float64(maxValues[i-1][0]+b[i]-1), float64(maxValues[i-1][1]+int(math.Abs(float64(b[i-1]-b[i]))))))
	}
	return int(math.Max(float64(maxValues[len(b)-1][0]), float64(maxValues[len(b)-1][1])))
}
